Add the downward monotonic cone to the cone catalogue

orlitzky opened the pull request:
This an an old PR that I never migrated from trac. We have a catalogue of convex cones accessible through `cones.foo()`, and this PR adds the downward monotonic cone to the catalogue. It will also close https://github.com/sagemath/sage/issues/30200.

orlitzky committed:
src/sage/geometry/cone_catalog.py: add the downward-monotonic cone

The downward-monotonic cone is the dual of the Schur cone that one
can obtain by running

  sage: cones.schur(n)

It is used in convex optimization, and is nice to have available
independent of the Schur cone, because cones.schur_cone(n).dual()
not only gives you back a cone that lives in the dual lattice of
the Schur cone, but is slower than constructing the downward-
monotonic cone directly. The new cone is available as

  sage: cones.downward_monotonic(n)

in the cone catalogue.

orlitzky committed:
src/doc: add the global Niez1998 reference

Add the paper "Group majorization and Schur type inequalities" by
Marek Niezgoda to src/doc/en/reference/references/index.rst, which
contains the global reference list. This paper mentions several
convex cones in the context of group majorization, and can be used
as a reference for some things in the cone catalogue.

Here's a BibTeX entry for good measure:

  @article{
    niez1998,
    author = "Marek Niezgoda",
    title = "Group majorization and {S}chur type inequalities",
    journal = "Linear Algebra and its Applications",
    volume = 268,
    number = 1,
    year = 1998,
    pages = "9--30",
    doi = "10.1016/S0024-3795(97)89322-6"
  }

orlitzky committed:
src/sage/geometry/cone_catalog.py: fix the majorization order

I have used the phrase "majorization order" in the cone catalogue in a
few places. It is however only a preorder, and not a partial order, so
the first order of business is to change that phrase.

There was also one test case that used a majorized_by() function whose
implementation was incorrect. No harm was caused by this, but as a
personal rule I prefer it if a function called majorized_by() actually
implements the majorization order. This commit fixes that as well, by
rearranging the arguments' elements in nonincreasing order, so that
the function truly tests whether or not one argument majorizes the
other.

orlitzky committed:
src/sage/geometry/cone_catalog.py: disambiguate a :class: reference

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -554,21 +705,25 @@ def schur(ambient_dim=None, lattice=None):
         sage: x = V.random_element()
         sage: y = V.random_element()
-        sage: majorized_by(x,y) == ( (y-x) in S )
+        sage: majorized_by(x,y) == ((rearrange(y)-rearrange(x)) in S)
```
```suggestion
        sage: majorized_by(x,y) == (rearrange(y) - rearrange(x) in S)
```

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -151,6 +161,143 @@ def _preprocess_args(ambient_dim, lattice):
+    from sage.matrix.constructor import matrix
+    from sage.rings.integer_ring import ZZ
+
+    (ambient_dim, lattice) = _preprocess_args(ambient_dim, lattice)
```
```suggestion
    ambient_dim, lattice = _preprocess_args(ambient_dim, lattice)
```
just a suggestion

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -151,6 +161,143 @@ def _preprocess_args(ambient_dim, lattice):
+    # Example 2.2 if you don't want to compute them yourself.
+    G = matrix.identity(ZZ, ambient_dim)
+    for i in range(1, ambient_dim):
+        G.add_multiple_of_row(i, i-1, 1)
```
```suggestion
        G.add_multiple_of_row(i, i - 1, 1)
```

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -151,6 +161,143 @@ def _preprocess_args(ambient_dim, lattice):
+    default lattice of rank ``ambient_dim`` will be used.
+
+    A ``ValueError`` is raised if neither ``ambient_dim`` nor
+    ``lattice`` are specified. It is also a ``ValueError`` to specify
```
```suggestion
    A :class:`ValueError` is raised if neither ``ambient_dim`` nor
    ``lattice`` are specified. It is also a :class:`ValueError` to specify
```

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -151,6 +161,143 @@ def _preprocess_args(ambient_dim, lattice):
+    nonincreasing order. Each generating ray has the integer ring as
+    its base ring.
+
+    A ``ValueError`` can be raised if the inputs are incompatible or
```
```suggestion
    A :class:`ValueError` can be raised if the inputs are incompatible or
```

mkoeppe commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -151,6 +161,143 @@ def _preprocess_args(ambient_dim, lattice):
+        ....:      for x in K.rays() )
+        True
+        sage: x = K.random_element()
+        sage: all( x[i] >= x[i+1] for i in range(ambient_dim - 1) )
```
```suggestion
        sage: all(x[i] >= x[i + 1]
        ....:      for i in range(ambient_dim - 1)
        ....:      for x in K.rays())
        True
        sage: x = K.random_element()
        sage: all(x[i] >= x[i + 1] for i in range(ambient_dim - 1))
```

mkoeppe reviewed (commented):
Otherwise LGTM

orlitzky committed:
src/sage/geometry/cone_catalog.py: add spaces around + and -

Thanks to Matthias Koeppe for the suggestion.

orlitzky committed:
src/sage/geometry/cone_catalog.py: ``ValueError`` -> :class:`ValueError`

...and rewrap the docstrings afterwards. Thanks to Matthias Koeppe for
the suggestion.

orlitzky committed:
src/sage/geometry/cone_catalog.py: drop parens from tuple assignments

For example,

  (x,y) = foo

becomes

  x,y = foo

Thanks to Matthias Koeppe for the suggestion.

orlitzky committed:
src/sage/geometry/cone_catalog.py: downward monotonic -> downward monotone

Rename this cone. "Monotone" is more popular than "monotonic", but
once you include the upward- or downward- prefix, "monotonic" begins
to win. I like the bike shed just the way it is, but the color isn't
important enough to me to leave the bike outside for another four
years while we ponder it.
